### PR TITLE
feat(myprofile): feature/icky 661 my account my profile

### PR DESCRIPTION
**Pull request by kibo-Kripa**
My account my profile. 
User can edit and save name, email id and password

### CODE REVIEW

File: components/my account/my profile/MyProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+  const { user, isAuthenticated } = useAuthContext()
+  const { t } = useTranslation(['checkout', 'common'])
+  const { updateUserData } = useUpateUserMutations()
+  const { updateUserPasswordData } = useUpdateUserPasswordMutations()
+  const [edit, setEdit] = useState(false)
+  const [editName, setEditName] = useState(false)
+  const [editEmail, setEditEmail] = useState(false)
+  const [editPassword, setEditPassword] = useState(false)
+  const [showPassword, setShowPassword] = useState<boolean>(false)
+  const [currentUser, setCurrentUser] = useState({
```
Comment (kibo-kevinwatts) [inline components/my account/my profile/MyProfile.tsx:25]: nothing in here should set the 'currentUser' , if this is for "formData" please update the name

File: components/my account/my profile/MyProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+      setCurrentUser({
+        id: user?.id as number,
+        firstName: user?.firstName as string,
+        lastName: user?.lastName as string,
+        emailAddress: user?.emailAddress as string,
+      })
+    }
+  }, [user])
+
+  interface disable {
```
Comment (kibo-kevinwatts) [inline components/my account/my profile/MyProfile.tsx:69]: don't define typescript types within component, should be at top of the file

File: components/my account/my profile/MyProfile.tsx [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+
+  useEffect(() => {
+    if (
+      currentUser.firstName !== user?.firstName ||
+      currentUser.lastName !== user?.lastName ||
+      currentUser.emailAddress !== user?.emailAddress
+    ) {
+      setDisableButton({ forName: false, forEmail: false })
+    } else {
+      setDisableButton({ forName: true, forEmail: true })
```
Comment (kibo-kevinwatts) [inline components/my account/my profile/MyProfile.tsx:87]: We need more validation than this, check other forms, 'yup' for examples

File: components/my account/my profile/MyProfile.tsx [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    }
+  }, [currentUser])
+
+  const {
+    formState: { errors },
+    control,
+    watch,
+  } = useForm({
```
Comment (kibo-kevinwatts) [inline components/my account/my profile/MyProfile.tsx:95]: Take a look at other forms that use, "yup" for validation.
Also want to make use of "isValid" to check form state is valid and the "handleSubmit" methods that ensure the form is valid before taking some action.
Currently, your handlers (at least for password) are doing no validation

**Review by kibo-kevinwatts — CHANGES_REQUESTED** [review]

File: components/my account/my profile/MyProfile.tsx [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+              <Box sx={style.box3}>
+                <Typography variant="body2"> {t('common:edit')}</Typography>
+              </Box>
+            </Box>
+            <Box sx={style.box2}>
+              <Box>
+                <Typography variant="body2" sx={{ fontWeight: 'bold' }}>
+                  {t('checkout:password')}
+                </Typography>
+                <Password fontSize='small'></Password>
```
Comment (kibo-chandradeeptalaha) [inline components/my account/my profile/MyProfile.tsx:442]: use "**********" instead of icon.

**Review by kibo-chandradeeptalaha — CHANGES_REQUESTED** [review]
Rebase the code. Remove the paading of the myProfile section.
Initial render doesn't show firstname/email.
Replace the password icon for display value with string "*************"

File: hooks/mutations/useProfile/useChangeCustomerData/useUpdateCustomerData.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  const response = await client.request({
+    document: updateCustomerData,
+    variables,
+  })
+
+  return response
+}
+
+export const useUpateUserMutations = () => {
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply hooks/mutations/useProfile/useChangeCustomerData/useUpdateCustomerData.ts:33]: resolved

File: components/my account/my profile/MyProfile.tsx [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+    lastName: '',
+    emailAddress: '',
+    id: 0,
+  })
+  const [updatedPassword, setUpdatedpassword] = useState({
+    oldPassword: '',
+    newPassword: '',
+    externalPassword: '',
+  })
+  const style = {
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:36]: resolved

File: hooks/mutations/useProfile/useUpdateCustomerPassword/useUpdatePassword.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+export const useUpdateUserPasswordMutations = () => {
+  const queryClient = useQueryClient()
+  return {
+    updateUserPasswordData: useMutation(updateUserPassword, {
+      onSuccess: () => {
+        queryClient.invalidateQueries(UpdatePasswordKeys.all)
+      },
+    }),
+  }
+}
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply hooks/mutations/useProfile/useUpdateCustomerPassword/useUpdatePassword.ts:41]: Done

File: components/my account/my profile/MyProfile.tsx [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+                  >
+                    {t('common:save')}
+                  </Button>
+                </Box>
+              </>
+            )}
+            {editPassword && (
+              <>
+                <Typography variant="h4" sx={{ fontWeight: 'bold' }}>
+                  {editPassword ? 'Edit Password' : 'Password'}
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:288]: Done

File: components/my account/my profile/MyProfile.tsx [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+                  >
+                    {t('common:save')}
+                  </Button>
+                </Box>
+              </>
+            )}
+            {editEmail && (
+              <>
+                <Typography variant="h4" sx={{ fontWeight: 'bold' }}>
+                  {editEmail ? 'Edit Email' : 'Email'}
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:233]: Resolved

File: components/my account/my profile/MyProfile.tsx [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+
+  return (
+    <Box>
+      <Box sx={style.box1} style={{ width: mobileView ? '100%' : '50%' }}>
+        {edit ? (
+          <>
+            {editName && (
+              <>
+                <Typography variant="h4" sx={{ fontWeight: 'bold' }}>
+                  {editName ? 'Edit Customer Name' : 'Customer Name'}
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:155]: resolved

File: components/my account/my profile/MyProfile.tsx (hunk 2 above)
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:69]: resolved

File: components/my account/my profile/MyProfile.tsx [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+
+import { Password, Visibility, VisibilityOff } from '@mui/icons-material'
+import { Box, Button, Typography, useMediaQuery, useTheme } from '@mui/material'
+import { useTranslation } from 'next-i18next'
+import { Controller, useForm } from 'react-hook-form'
+
+import { KiboTextBox ,PasswordValidation} from '@/components/common'
+import { useAuthContext } from '@/context'
+import { useUpateUserMutations } from '@/hooks/mutations/useProfile/useChangeCustomerData/useUpdateCustomerData'
+import { useUpdateUserPasswordMutations } from '@/hooks/mutations/useProfile/useUpdateCustomerPassword/useUpdatePassword'
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:13]: Done

File: components/my account/my profile/MyProfile.tsx [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+
+  interface disable {
+    forName: boolean
+    forEmail: boolean
+  }
+
+  const [disableButton, setDisableButton] = useState<disable>({
+    forName: true,
+    forEmail: true,
+  })
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:77]: Done

File: components/my account/my profile/MyProfile.tsx [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+                    )}
+                  />
+                </Box>
+                <Box
+                  sx={{
+                    display: 'flex',
+                    gap: 1,
+                    width: '50%',
+                    flexDirection: 'column',
+                  }}
```
[the remark this answers is not in the document]
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:207]: Done

File: components/my account/my profile/MyProfile.tsx (hunk 5 above)
Comment (kibo-Kripa) [reply components/my account/my profile/MyProfile.tsx:442]: done

File: components/my-account/MyProfile/MyProfile.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+describe('[components] MyProfile', () => {
+  it('should render component', () => {
+    setup()
+
+    const customerNameText = screen.getByText(/customer-name/i)
+    const emailText = screen.getByText(/email/i)
+    const passwordText = screen.getByText(/password/i)
+    const editName = screen.getAllByText(/edit/i)
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/MyProfile/MyProfile.spec.tsx:37]: change editName to allEditLinks.

change all editName variables to editLink(singular) / allEditLinks(plural)

File: components/my-account/MyProfile/MyProfile.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+    expect(customerNameText).toBeVisible()
+    expect(emailText).toBeVisible()
+    expect(passwordText).toBeVisible()
+    expect(editName[0]).toBeVisible()
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/MyProfile/MyProfile.spec.tsx:42]: assert if all the edit links are visible or not. No only the first one

File: components/my-account/MyProfile/MyProfile.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+  render(<Common {...Common.args} />)
+
+  return {
+    user,
+  }
+}
+
+describe('[components] MyProfile', () => {
+  it('should render component', () => {
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/MyProfile/MyProfile.spec.tsx:31]: Test the passed values from story are visible or not.

File: components/my-account/MyProfile/MyProfile.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+import { composeStories } from '@storybook/testing-react'
+import { render, screen } from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+
+import * as stories from './MyProfile.stories'
+
+const { Common } = composeStories(stories)
+
+const onChangMock = jest.fn()
+const KiboTextBoxMock = () => <input data-testid="text-box-mock" onChange={onChangMock} />
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/MyProfile/MyProfile.spec.tsx:12]: Don't mock these. Try to test if firstName/Lastname textbox is visible or not if we click on the respective edit link. 
Do these for all

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx [hunk 19]
```diff
@@ -0,0 +1,298 @@
+import React, { useEffect, useState } from 'react'
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx:1]: Add localization for strings that will be visible on the UI. eg: for error messages defined inside schema

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx [hunk 20]
```diff
@@ -0,0 +1,298 @@
...
+
+import { yupResolver } from '@hookform/resolvers/yup'
+import { Visibility, VisibilityOff } from '@mui/icons-material'
+import { styled, FormControl, Box, Stack, Button } from '@mui/material'
+import { useTranslation } from 'next-i18next'
+import { useForm, Controller } from 'react-hook-form'
+import * as yup from 'yup'
+
+import { PasswordValidation } from '@/components/common'
+import KiboTextBox from '@/components/common/KiboTextBox/KiboTextBox'
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx:11]: change to 

import { PasswordValidation, KiboTextBox  } from '@/components/common'

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.stories.tsx [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+)
+
+// Common
+export const Common = Template.bind({})
+
+Common.args = {
+  firstName: '',
+  lastName: '',
+  emailAddress: '',
+  isEmailForm: false,
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.stories.tsx:23]: isPasswordForm missing

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.stories.tsx [hunk 22]
```diff
@@ -0,0 +1,24 @@
+import React from 'react'
+
+import { ComponentStory, ComponentMeta } from '@storybook/react'
+
+import ProfileDetailsForm from './ProfileDetailsForm'
+
+export default {
+  component: ProfileDetailsForm,
+  title: 'My Profile/ProfileDetailsForm',
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.stories.tsx:9]: add argTypes of 
onSaveProfileData , onCancel

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+import { composeStories } from '@storybook/testing-react'
+import { render, screen } from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+
+import * as stories from './ProfileDetailsForm.stories' // import all stories from the stories file
+
+const { Common } = composeStories(stories)
+
+const KiboTextBoxMock = () => (
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.spec.tsx:11]: Don't mock. Try to test using the props. if isEmailForm is true, does the email textbox appears ? 
Try to do like this for all

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+describe('[component] - ProfileDetailsForm', () => {
+  const setup = () => {
+    const user = userEvent.setup()
+    render(<Common {...Common.args} />)
+    return {
+      user,
+    }
+  }
+  it('should render component', () => {
```
Comment (kibo-chandradeeptalaha) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.spec.tsx:26]: More testcases required depending on the props passed

File: hooks/mutations/useProfile/useUpdateUserData/useUpdateUserData.ts [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+export interface UpdateUserDataInput {
+  id: number
+  firstName: string
+  lastName: string
+  emailAddress: string
+}
+
+interface UpdateUserDataProps {
+  accountId: number
+  customerAccountInput: UpdateUserDataInput
```
Comment (kibo-chandradeeptalaha) [inline hooks/mutations/useProfile/useUpdateUserData/useUpdateUserData.ts:16]: We do have a type for CustomerAccountInput no ? Can you check ?

File: hooks/mutations/useProfile/useUpdateUserPassword/useUpdateUserPassword.ts [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+
+export interface UpdateUserPasswordInput {
+  oldPassword: string
+  newPassword: string
+  externalPassword: string
+}
+
+interface UpdateUserPasswordProps {
+  accountId: number
+  passwordInfoInput: UpdateUserPasswordInput
```
Comment (kibo-chandradeeptalaha) [inline hooks/mutations/useProfile/useUpdateUserPassword/useUpdateUserPassword.ts:15]: check if we have types for passwordInfoInput

**Review by kibo-chandradeeptalaha — CHANGES_REQUESTED** [review]
Please take a look

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx [hunk 27]
```diff
@@ -0,0 +1,299 @@
...
+    defaultValues: getDefaultValues(),
+    criteriaMode: 'firstError',
+    resolver: yupResolver(cardSchema),
+    shouldFocusError: true,
+    context: { isEmailForm, isPasswordForm },
+  })
+  const userEnteredPassword = watch('newPassword')
+  const confirmationPassword = watch('confirmPassword')
+
+  const buildParam = (formData: ProfileDetails) => {
```
Comment (kibo-kevinwatts) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx:121]: Should this function be in the component and not in the hook?
its mostly just returning the same object anyway?

File: hooks/mutations/useProfile/useUpdateUserPassword/useUpdateUserPassword.ts [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+
+  return response
+}
+
+export const useUpdateUserPasswordMutations = () => {
+  const queryClient = useQueryClient()
+  return {
+    updateUserPasswordData: useMutation(updateUserPassword, {
+      onSuccess: () => {
+        queryClient.invalidateQueries(loginKeys.user)
```
Comment (kibo-kevinwatts) [inline hooks/mutations/useProfile/useUpdateUserPassword/useUpdateUserPassword.ts:32]: this isn't necessary

File: test-report.xml [hunk 29]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<testExecutions version="1">
```
Comment (kibo-kevinwatts) [inline test-report.xml:2]: delete this and update git ignore to ignore report file

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+      }
+    }),
+    currentPassword: yup.string().when('$isPasswordForm', (isPasswordForm, schema) => {
+      if (isPasswordForm) {
+        return schema.required(t('this-field-is-required'))
+      }
+    }),
+    newPassword: yup.string().when('$isPasswordForm', (isPasswordForm, schema) => {
+      if (isPasswordForm) {
+        return schema
```
Comment (kibo-kevinwatts) [inline components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx:49]: Is this the same way we are doing password validation in other forms?
@kibo-sushant ?

File: components/my-account/MyProfile/MyProfile.tsx [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+    },
+    {
+      id: ProfileSections.Password,
+      label: t('password'),
+      value: '**************',
+    },
+  ]
+
+  useEffect(() => {
+    setCurrentUser({
```
Comment (kibo-kevinwatts) [inline components/my-account/MyProfile/MyProfile.tsx:61]: What is this "currentUser" in this component for?

**Review by kibo-kevinwatts — CHANGES_REQUESTED** [review]

File: components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx (hunk 27 above)
Comment (kibo-chandradeeptalaha) [reply components/my-account/ProfileDetailsForm/ProfileDetailsForm.tsx:121]: I created this to pass the exact object based on the form type to the parent component.

File: components/my-account/MyProfile/MyProfile.tsx (hunk 31 above)
Comment (kibo-chandradeeptalaha) [reply components/my-account/MyProfile/MyProfile.tsx:61]: Removed it.

File: test-report.xml (hunk 29 above)
Comment (kibo-chandradeeptalaha) [reply test-report.xml:2]: Done

File: hooks/mutations/useProfile/useUpdateUserPassword/useUpdateUserPassword.ts (hunk 28 above)
Comment (kibo-chandradeeptalaha) [reply hooks/mutations/useProfile/useUpdateUserPassword/useUpdateUserPassword.ts:32]: Removed

File: components/my-account/MyProfile/MyProfile.tsx [hunk 32]
```diff
@@ -0,0 +1,142 @@
...
+        },
+      })
+    } catch (err) {
+      console.log(err)
+    }
+    setCurrentEditableField(null)
+  }
+
+  const handleUpdateUserPassword = async (updatedPassword: PasswordTypes) => {
+    try {
```
Comment (amolpawar-kibo) [inline components/my-account/MyProfile/MyProfile.tsx:76]: could you please destructure props & use it?
ex. const {oldPassword} = updatedPassword

File: components/my-account/MyProfile/MyProfile.tsx [hunk 33]
```diff
@@ -0,0 +1,142 @@
...
+      handleUpdateUserPassword(profileFormData)
+      return
+    }
+
+    try {
+      await updateUserData.mutateAsync({
+        accountId: id,
+        customerAccountInput: {
+          id: id,
+          firstName: profileFormData.firstName || firstName,
```
Comment (amolpawar-kibo) [inline components/my-account/MyProfile/MyProfile.tsx:64]: could you please destructure props & use it?

File: components/my-account/MyProfile/MyProfile.tsx [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+                  {each.label}
+                </Typography>
+                <Typography variant="body1" data-testid={each.label}>
+                  {each.value}
+                </Typography>
+              </Stack>
+              <Typography
+                variant="body1"
+                onClick={() => {
+                  setCurrentEditableField(each.id)
```
Comment (amolpawar-kibo) [inline components/my-account/MyProfile/MyProfile.tsx:128]: Could you please rewrite onClick={() => {
                  setCurrentEditableField(each.id)} to onClick={() => setCurrentEditableField(each.id)}

**Review by kibo-kevinwatts — APPROVED** [review]

**Comment by kibo-kevinwatts** [comment]
lets resolve the conflicts and get this merged

**Review by TriptiVermaXP — APPROVED** [review]

**Pull request merged**